packages/alpinejs/src/evaluator.js

```diff
@@ -117,7 +103,7 @@ function generateEvaluatorFromString(dataStack, expression, el) {
             // Check if the function ran synchronously,
             if (func.finished) {
                 // Return the immediate result.
-                runIfTypeOfFunction(receiver, func.result, completeScope, params, el)
+                runIfTypeOfFunction(autoEvaluateFunctions, receiver, func.result, completeScope, params, el)
```

Comment:
Is there a clean way to avoid calling this entirely if we aren't auto evaluating functions?

Since why call `runIfTypeOfFunction` if we aren't auto evaluating it, right? Then the function name doesn't make much sense.

Reply:
You mean to conditionally execute it if it's a function, else simply return the value?

One thing to mind is that `runIfTypeOfFunction` will execute the function once, and return whatever value it is (including another function), but if it returns another `async` function, then it will continue executing it recursively. In other words, when it's `async`, it will unwrap all the promises and return the final value.

Check this example.

```html
<html>
<head>
    <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
</head>
<body>
    <h1 x-data="{ fn: () => () => () => () => 5 }" x-text="fn"></h1>
    <h1 x-data="{ fn: async () => async () => async () => async () => 5 }" x-text="fn"></h1>
</body>
</html>
```

Result on screen:

```
() => () => () => 5
5
```

So `runIfTypeOfFunction` not only changes with different values of the auto-execute flag, but also whether the values are promises or not. I don't feel confident enough to decide how to refactor this, because maybe some part of the code needs the recursive unwrapping (although I've never seen something like this in the source code). I'd personally just leave it like that and refactor it next time, which wouldn't be that hard considering `runIfTypeOfFunction` is contained (private) inside `evaluator.js` (alpinejs/csp).

Reply:
I didn't write it in the first place, but I feel I can say with high confidence that the case of functions returning promises that resolve to functions that return promises that resolve to functions was never a case that came up when writing that code...

Reply:
@ekwoka I refactored and renamed the function.

* Refactored runIfTypeOfFunction to remove the recursive behavior. Now the behavior is more predictable: it just evaluates once without recursively unwrapping functions (which by the way only worked for promises, which made it even harder to understand what it was doing)
* Renamed runIfTypeOfFunction to handleEvalResult, because the name may not make sense anymore. Basically the purpose of this function is to send a result to the receiver, so this name is more generic I guess.

Code snippet:

```js
export function handleEvalResult(autoEvaluateFunctions, receiver, value, scope, params, el) {
    const shouldCallFunc = autoEvaluateFunctions && typeof value === 'function'

    const result = shouldCallFunc ? value.apply(scope, params) : value

    if (result instanceof Promise) {
        result.then(i => receiver(i)).catch( error => handleError( error, el, value ) )
    } else {
        receiver(result)
    }
}
```

Reply:
That's looking a lot better.

But I think we go further!

What if being passed to handle eval result, the autoEvaluateFunctions prop just wrapped the `receiver`? 🤔

Reply:
like

```js

const inner receiver = autoCall 
  ? (maybeFn) => maybeFn instanceof Function 
    ? receiver(maybeFn.apply(scope, params)) 
    : receiver(maybeFn) 
  : receiver

```

well, prettier than that I mean...

Reply:
@ekwoka Thank you for the clarification! I think the problem is that this would make other codes more verbose. The wrapping logic would need to be done in the caller, and currently there are several places where `handleEvalResult(...)` is called, so all those places would need to get modified.

Basically all places that look like this (three in total):

```js
    return (receiver = () => {}, /* ... */) => {
        // ....

        handleEvalResult(autoEvaluateFunctions, receiver, result)
    }
```

So I guess in this case it'd be better to just leave `handleEvalResult` do the dirty job of deciding whether to execute or not. What do you think?

Reply:
@ekwoka Yes, I think it's a bit difficult to abstract it further (without doing a major revamp)

So from me that's all I have to add to the PR, it achieves everything I wanted, with only exporting `injectMagics` and by refactoring the other involved parts (`handleEvalResult` function)

Reply:
👍